### DISCUSSION TITLE
Is zlib CVE-2023-45853 exploitable in PyInstaller?

### DISCUSSION THREAD

**Discussion by mmirkows** [pyinstaller/pyinstaller#8064] [Help] [answered]
Hi,
there is a critical CVE in zlib https://nvd.nist.gov/vuln/detail/CVE-2023-45853 (no fixed version released yet). 
Do you know if PyInstaller is exploitable or not affected?

Regards,
Michal Mirkowski

**Comment by rokm** [comment, marked as answer]
We don't use the MiniZip part of `zlib` anywhere, so no.

**Comment by raghupcr** [comment]
Hi Team,

we have also got critical vulnerability on Pyinstaller which internally uses zlib. current version is zlib 1.2.13 but there is vulnerability in this version and need to upgrade to zlib 1.3.1. Will it be upgraded by any chance or based on the above comments will that be ignored as it is not using minizip? 

Please provide your input

Thanks,
Raghavendra

**Reply by bwoodsend** [reply]
Vulnerabilities in zlib are irrelevant to PyInstaller because PyInstaller does not use zlib on untrusted inputs.

**Reply by raghupcr** [reply]
Thanks @bwoodsend but the last time I saw it was upgraded from 1.2.11 to 1.2.13, which helped us to resolve one security issue. We were hoping that the new issue would also get resolved if Pyinstaller moved to the latest 1.3.1 version. If not for the security issue are there any plans to move to the latest ZLIB version? any timeline?

**Reply by bwoodsend** [reply]
We upgrade it on a whim every now and then it's never a security update – just avoiding letting dependencies rot too much. Certainly, I'm not going to bump it and go through a new release just to appease whatever scanner your using to detect this false positive warning.

**Reply by raghupcr** [reply]
Thanks @bwoodsend for the input, we will check with our Security team internally what can be done as there will be no release with this CVE fix.

**Reply by raghupcr** [reply]
@bwoodsend one more input required from your side, can we take the pyinstaller code and build to the custom zlib version? If we can do that will that be fine, what are all the things we need to take care of if we want to do that custom building?

**Reply by rokm** [reply]
> @bwoodsend one more input required from your side, can we take the pyinstaller code and build to the custom zlib version? If we can do that will that be fine, what are all the things we need to take care of if we want to do that custom building?

Nothing should prevent you from updating bundled zlib code and [rebuilding the bootloader](https://pyinstaller.org/en/stable/bootloader-building.html) yourself.

**Reply by raghupcr** [reply]
@rokm Thanks for the update, so on windows pyinstaller will use the zlib which is present in the bootloader so if we change that zlib to a different version then it will solve the vulnerability which our tool is showing as of now. One more thing is there a option for pyinstaller not using the bootloader zlib code and use the custom zlib code which is present on the windows machine? on linux it is possible on windows is it possible?

**Reply by bwoodsend** [reply]
> One more thing is there a option for pyinstaller not using the bootloader zlib code and use the custom zlib code which is present on the windows machine?

Windows doesn't provide a copy of zlib so not if you want it to work.

**Reply by rokm** [reply]
On Windows, bootloader is built with a static copy of zlib, which is built from bundled sources (found in `bootloader/zlib`). So if you want to have the latest zlib version, you will need to update the bundled sources, and rebuild the bootloader. We do not provide an option for dynamic-linked zlib on Windows, because that is typically not available at run-time (so you would need to distribute the zlib DLL together with executable, which beats the purpose of onefile builds).

**Reply by raghupcr** [reply]
@rokm @bwoodsend  thank you, so the best option to resolve this from our side is to replace the zlib in the bootloader and build the pyinstaller again, and use it for building the application. please correct me if my understanding is wrong

**Reply by rokm** [reply]
> @rokm @bwoodsend thank you, so the best option to resolve this from our side is to replace the zlib in the bootloader and build the pyinstaller again, and use it for building the application. please correct me if my understanding is wrong

That's correct.

**Reply by rokm** [reply]
You can also grab https://github.com/pyinstaller/pyinstaller/pull/8292 now, but you will still need to rebuild the bootloader yourself.

**Reply by raghupcr** [reply]
@rokm #8292 Nice, This is what is required for us, so when this PR/MR will be part of pyinstaller? we can wait for that for some time so that zlib 1.3.1 will be part of pyinstaller

**Reply by raghupcr** [reply]
@rokm  with this #8292 is there a plan for the pyinstaller version build? can we just get the high level timeline when it will be released, so that we can plan accordingly

**Reply by rokm** [reply]
There will eventually be a new release, but there is no official timeline; if you want rough estimate, it might take more than a month, but probably not more than two months. If you need that zlib update sooner, you should update and rebuild the bootloader yourself.

**Reply by raghupcr** [reply]
@rokm Great thanks a lot for the information, this will help, as mentioned we are working in parallel with the bootloader build as well Thanks a lot

**Reply by bwoodsend** [reply]
I really would prefer you to just add an exclusion to whatever tool you're using. There will always be another irrelevant zlib vulnerability and I don't want people asking me for new releases every time it happens.

**Reply by raghupcr** [reply]
@bwoodsend Sure we are checking with our security team as well and trying to check if it is a concern from a release point of view. Anyway as mentioned by @rokm we are also looking into the bootloader build with specific zlib code which also helps if there are no pyinstaller releases planned for a specific vulnerability. Thanks for the inputs we will look into that aspect as well

**Reply by bwoodsend** [reply]
PyInstaller 6.4 is out now with the new zlib version. But like I said before, we will not update it every time a new irrelevant vulnerability comes floating by. If your security scanning tool doesn't let you easily exclude false positives or your security team or too dysfunctional to do it then you should seriously reconsider using this tool. Either way, it's your problem and I don't want to even hear about the next one – let alone be asked when it's getting fixed.

**Reply by raghupcr** [reply]
@rokm @bwoodsend thanks for the update. One query after upgrading to the latest pyinstaller 6.4 we are seeing only 2 files, one exe and one _internal folder where all the dependent files are kept. This was bit different from the previous pyinstaller. Is it ok if we move all the contents from _internal folder one step back where the exe is available?

**Reply by bwoodsend** [reply]
That switch came in v6.0. You can use `--contents-directory=.` to get the old behavior but there should be no good reason to want to do that.

**Reply by raghupcr** [reply]
@bwoodsend hi , when trying to use this command we are facing some issues as shown below
PS C:\GitLab-Runner\builds\f1oNDx7T\0\business\services\sdst\prodeploy\ese\ese> pyinstaller --clean -y --log-level DEBUG --contents-directory=. pyinstaller_ese_main.spec
1759 INFO: PyInstaller: 6.4.0, contrib hooks: 2024.1
1759 INFO: Python: 3.12.1
1941 INFO: Platform: Windows-10-10.0.19045-SP0
option(s) not allowed:
  --contents-directory
makespec options not valid when a .spec file is given

Is there anything wrong here ? we tried giving another option but same error, can you please provide some input on this?
PS C:\GitLab-Runner\builds\f1oNDx7T\0\business\services\sdst\prodeploy\ese\ese> pyinstaller --clean -y --log-level DEBUG --contents-directory dist/ese/ --specpath . pyinstaller_ese_main.spec
1767 INFO: PyInstaller: 6.4.0, contrib hooks: 2024.1
1775 INFO: Python: 3.12.1
1950 INFO: Platform: Windows-10-10.0.19045-SP0
option(s) not allowed:
  --specpath
  --contents-directory
makespec options not valid when a .spec file is given

Thanks,

**Reply by rokm** [reply]
When you are building from .spec file, you cannot specify options that are used for spec file generation (because you already have a generated spec). So you should add `contents_directory='.'` keyword argument to `EXE()` in the .spec file instead of specifying it on command line.

**Reply by raghupcr** [reply]
Thanks @rokm with that change we can see all the contents and the exe in the same folder which was like 5.x behavior. This is what is required as of now for us because of the immediate release this month. Thanks again.

one more query here we are observing a lot of folders in the newer pyinstaller version. A lot of folders are present as shown in the files.

**pyinstaller6 4**

![pyinstaller6 4](https://github.com/pyinstaller/pyinstaller/assets/100952677/8a72c126-c89d-4e4d-800c-482ff14aeae3)

**Pyinstaller 5.x**

![pyinstaller5](https://github.com/pyinstaller/pyinstaller/assets/100952677/cdabd438-f8da-438d-99b3-faab4531ddc3)

Is this expected behavior we have not changed anything in the spec file and in the dependent files. are there any options that are missing again? **Our executable size previously was 8MB now it became 6MB.** Can you please help with this

Snippet of the spec file

main_pyz = PYZ (
    main_service.pure,
    main_service.zipped_data,
    cipher=block_cipher
)

main_exe = EXE (
    main_pyz,
    main_service.scripts,
    [],
    exclude_binaries = True,
    name = main_name,
    debug = False,
    bootloader_ignore_signals = False,
    strip = False,
    upx = True,
    console = True,
    contents_directory = '.'
)

collect= COLLECT (
    main_exe,
    main_service.binaries,
    main_service.zipfiles,
    main_service.datas,
    strip = False,
    upx = True,
    name = main_name
)

**Reply by rokm** [reply]
> Is this expected behavior we have not changed anything in the spec file and in the dependent files

If nothing else, you are building with at least slightly different environment - e.g., version of cryptography package changed from 41.0.3 to 41.0.7, as seen from collected metadata directory. So the question is, what else has changed in the environment? Do you have more packages (e.g., some optional/conditionally-imported dependencies) installed?

> Snippet of the spec file

The part of .spec that might contain any clues would be the initial one - i.e., from the beginning to instantiation of `Analysis`.

**Reply by raghupcr** [reply]
@rokm yes, the environment is different one is build env and the other one is test environment where I have installed the previous build which is through pyinstaller 5.x version. As for your question the cryptography package has only changed. what we could see difference is al the packages have -dist-info in the screenshot shared

**Reply by rokm** [reply]
> @rokm yes, the environment is different one is build env and the other one is test environment where I have installed the previous build which is through pyinstaller 5.x version.

I mean, was version built with PyInstaller 6.4 built in an identical environment to the one where PyInstaller 5.x was used?

> what we could see difference is al the packages have -dist-info in the screenshot shared

Yes, more metadata was collected with 6.x. We don't automatically collect those - but if you aren't explicitly collecting them via `copy_metadata` at beginning of your spec, then maybe an updated version of hook (from `pyinstaller-hooks-contrib` is collecting those.

What is the output of `pip freeze`, and how does the beginning of your spec file (or rather, whole spec file) look like?

**Reply by raghupcr** [reply]
@rokm Yes spec file is same and the environment is same where pyinstaller 5.x was there and later 6.x got installed

complete spec file

```
import os


def fix_relative_root_path(path, *paths):
    return os.path.abspath(os.path.join(BASE_PATH, path, *paths))


# Needed for bundling obfuscated scripts
def patch_scripts(scripts, _src, _target):
    for i in range(len(scripts)):
        if scripts[i][1].startswith(_src):
            script_path = scripts[i][1].replace(_src, _target)
            if os.path.exists(script_path):
                scripts[i] = scripts[i][0], script_path, scripts[i][2]

def patch_modules(pure_modules, _src, _target):
    for i in range(len(pure_modules)):
        if pure_modules[i][1].startswith(_src):
            module_path = pure_modules[i][1].replace(_src, _target)
            if os.path.exists(module_path):
                if hasattr(pure_modules, '_code_cache'):
                    with open(module_path) as f:
                        pure_modules._code_cache[pure_modules[i][0]] = compile(f.read(), pure_modules[i][1], 'exec')
                pure_modules[i] = pure_modules[i][0], module_path, pure_modules[i][2]


block_cipher = None
# pyinstaller doesn't work well with relative paths, so make sure everything is an absolute path
# SPECPATH is the path to the spec file.  Could be absolute or relative to cwd.  If in cwd and relateive, it may be empty.
BASE_PATH = os.path.abspath(SPECPATH if SPECPATH else os.getcwd()) # empty spec path means current directory.
common_pathex = [
    fix_relative_root_path('testcatcher', 'lib', 'pytestCatcher'),
    fix_relative_root_path('dist', 'obf', 'testcatcher', 'lib', 'pytestCatcher')
]
common_datas = [
  (
     os.path.join(fix_relative_root_path('dist', 'obf', 'testcatcher', 'lib', 'pyarmor_runtime_005098'), '*'),
        'pyarmor_runtime_005098'
    ),
]

common_hidden_imports = ['cheroot.ssl', 'cheroot.ssl.builtin', 'pytransform']
common_hooks_path = [ fix_relative_root_path('dist', 'obf', 'testcatcher', 'lib', 'pytestCatcher') ]
main_name = 'ese'
src_script_path = fix_relative_root_path('testcatcher', 'libexec', 'ese', '__main__.py')
main_scripts = [ fix_relative_root_path('testcatcher', 'lib', 'pytestCatcher', '__main__.py') ]

if os.name == 'nt':
    # append some additional windows platform specific datas
    additional_data = {
        'utilities/' : os.path.join(fix_relative_root_path('utilities'), '*.dll'),
        'setup/' : os.path.join(fix_relative_root_path('installer'), 'setup.*'),
        'configuration/' : os.path.join(fix_relative_root_path('installer'), 'DefaultProperties.json')
    }
    for target, source in additional_data.items():
        common_datas.append((source, target))

    # specify additional windows platform specific hidden packages
    for package in ['jaraco', 'win32timezone']:
        common_hidden_imports.append(package)

# Check if main script exists or not, if not then copy src script at main script location
if not os.path.exists(main_scripts[0]):
    os.system(f"copy {src_script_path} {main_scripts[0]}")
    print ("Main script copied : ", os.path.exists(main_scripts[0]))

main_service = Analysis (
    main_scripts,
    pathex = common_pathex,
    binaries = [],
    datas = common_datas,
    hiddenimports = common_hidden_imports,
    hookspath = common_hooks_path,
    runtime_hooks = [],
    excludes = [
        'docutils',
        'chardet'
    ],
    win_no_prefer_redirects = False,
    win_private_assemblies = False,
    cipher = block_cipher,
    noarchive = False
)

# Patch by PyArmor
_src = fix_relative_root_path('testcatcher')
_target = fix_relative_root_path('dist', 'obf', 'testcatcher')
patch_scripts(main_service.scripts, _src, _target)
patch_modules(main_service.pure, _src, _target)
# Patch end.

main_pyz = PYZ (
    main_service.pure,
    main_service.zipped_data,
    cipher=block_cipher
)

main_exe = EXE (
    main_pyz,
    main_service.scripts,
    [],
    exclude_binaries = True,
    name = main_name,
    debug = False,
    bootloader_ignore_signals = False,
    strip = False,
    upx = True,
    console = True,
    contents_directory = '.'
)

collect_ese = COLLECT (
    main_exe,
    main_service.binaries,
    main_service.zipfiles,
    main_service.datas,
    strip = False,
    upx = True,
    name = main_name
)
```

**Reply by rokm** [reply]
> @rokm Yes spec file is same and the environment is same where pyinstaller 5.x was there and later 6.x got installed

Do you still have the 5.x environment available? What is the output of `pip freeze` in environment with 6.4 and what is the output of `pip freeze` in environment with 5.x?

**Reply by rokm** [reply]
Ah, nevermind, I missed the part where you said you upgraded PyInstaller from 5.x to 6.x in the same environment.

In that case, can you build with `--log-level DEBUG` and attach the full build log (attach it as a text file here, or upload it somewhere so I can take a look at it). Perhaps that will be enough to tell us where the extra metadata came from...

**Reply by raghupcr** [reply]
@rokm  sorry for the delayed response. Discussed with the team and came to know that .dist-info files were getting generated in the pyinstaller 5.x also. After that as part of the InstallShield we are removing the folders it seems. This i came to know today only. So this is expected behavior. Thanks once again @rokm for your valuable inputs and time :) we are in the process of testing the output (windows final o/p based out of Installshield)
